cpp/src/arrow/compute/kernels/aggregate_basic.cc

```diff
@@ -255,6 +255,28 @@ struct SumState {
   }
 };
 
+template <>
+struct SumState<BooleanType> {
+  using SumType = typename FindAccumulatorType<BooleanType>::Type;
+  using ThisType = SumState<BooleanType, SumType>;
+
+  ThisType& operator+=(const ThisType& rhs) {
```

Comment:
I chose to specialize SumState since the required machinery is already implemented. 

I could factor out the common parts to a SumStateBase or something. Let me know if there is a nicer solution.